support/mesh/mesh_channel_core/src/mpsc.rs

```diff
@@ -0,0 +1,971 @@
+    /// another sender in the message you send so that the receiving thread can
+    /// use it to send a response.
+    ///
+    /// ```rust
+    /// # use mesh_channel_core::*;
+    /// # futures::executor::block_on(async {
+    /// let (send, mut recv) = channel();
+    /// let (response_send, mut response_recv) = channel::<bool>();
+    /// send.send((3, response_send));
+    /// let (val, response_send) = recv.recv().await.unwrap();
+    /// response_send.send(val == 3);
+    /// assert_eq!(response_recv.recv().await.unwrap(), true);
+    /// # });
+    /// ```
+    pub fn send(&self, message: T) {
+        let mut message = MaybeUninit::new(message);
+        // SAFETY: the queue is for `T` and `message` is a valid owned `T`.
+        // Additionally, the sender/receiver is only `Send`/`Sync`  if `T` is
+        // `Send`/`Sync`.
+        let sent = unsafe { self.0.send(MessagePtr::new(&mut message)) };
+        if !sent {
+            // SAFETY: `message` was not dropped.
+            unsafe { message.assume_init_drop() };
+        }
+    }
+
+    /// Returns whether the receiving side of the channel is known to be closed
+    /// (or failed).
+    ///
+    /// This is useful to determine if there is any point in sending more data
+    /// via this port. But even if this returns `false` messages may still fail
+    /// to reach the destination.
+    pub fn is_closed(&self) -> bool {
+        self.0.is_closed()
+    }
+}
+
+struct MessagePtr(*mut ());
+
+impl MessagePtr {
+    fn new<T>(message: &mut MaybeUninit<T>) -> Self {
+        Self(message.as_mut_ptr().cast())
+    }
+
+    /// # Safety
+    /// The caller must ensure that `self` is a valid owned `T`.
+    unsafe fn read<T>(self) -> T {
+        // SAFETY: The caller guarantees `self` is a valid owned `T`.
+        unsafe { self.0.cast::<T>().read() }
+    }
+}
+
+impl SenderCore {
+    /// Sends `message`, taking ownership of it.
+    ///
+    /// Returns `true` if the message was sent. If `false`, the caller retains
+    /// ownership of the message and must drop it.
+    ///
+    /// # Safety
+    /// The caller must ensure that the message is a valid owned `T` for the `T`
+    /// the queue was created with. It also must ensure that the queue is not
+    /// sent/shared across threads unless `T` is `Send`/`Sync`.
+    #[must_use]
+    unsafe fn send(&self, message: MessagePtr) -> bool {
+        match self.0.access() {
+            QueueAccess::Local(mut local) => {
+                if local.receiver_gone {
+                    return false;
+                }
+                // SAFETY: The caller guarantees `message` is a valid owned `T`,
+                // and that the queue will not be sent/shared across threads
+                // unless `T` is `Send`/`Sync`.
+                unsafe { local.messages.push_back(message.0) };
+                if let Some(waker) = local.waker.take() {
+                    drop(local);
```

Comment:
what's with the explicit drop here?

Reply:
In general it's best not to call wakers under locks. We use this pattern throughout the codebase.